qiskit/providers/__init__.py

```diff
@@ -110,6 +109,394 @@
+so that the :class:`~qiskit.transpiler.passes.BasisTranslator` used in the
+:func:`~qiskit.compiler.transpile` function and :mod:`~qiskit.transpiler`
+module can convert an arbitrary circuit using the custom basis set. This
+can be done by defining equivalent circuits, in terms of the custom gate,
+for standard gates. Typically if you can convert from a
+:class:`~qiskit.circuit.library.CXGate` (if your basis doesn't include a
+standard 2 qubit gate) and that some commonly used single
+qubit rotation gates like :class:`~qiskit.ciruit.library.HGate`. Typically
+the more equivalence rules that are defined from standard gates to your basis
+the more efficient translation from an arbitrary circuit to the target
+basis will be (although not always, and there is a diminishing margin of return)
+
+For example, if you were to add some rules for the above custom ``SYGate``
+we could define the :class:`~qiskit.circuit.library.U2Gate` and
+:class:`~qiskit.circuit.library.HGate`::
+
+    from qiskit.circuit.equivalence_library import SessionEquivalenceLibrary
+    from qiskit.circuit.library import HGate
+    from qiskit.circuit.library import ZGate
+    from qiskit.circuit.library import RZGate
+    from qiskit.circuit.library import U2Gate
+
+
+    # H => Z SY
+    q = qiskit.QuantumRegister(1, "q")
+    def_sy_h = qiskit.QuantumCircuit(q)
+    def_sy_h.append(ZGate(), [q[0]], [])
+    def_sy_h.append(SYGate(), [q[0]], [])
+    SessionEquivalenceLibrary.add_equivalence(
+        HGate(), def_sy_h)
+
+    # u2 => Z SY Z
+    phi = qiskit.circuit.Parameter('phi')
+    lam = qiskit.circuit.Parameter('lambda')
+    q = qiskit.QuantumRegister(1, "q")
+    def_sy_u2 = qiskit.QuantumCircuit(q)
+    def_sy_u2.append(RZGate(lam), [q[0]], [])
+    def_sy_u2.append(SYGate(), [q[0]], [])
+    def_sy_u2.append(RZGate(phi), [q[0]], [])
+    SessionEquivalenceLibrary.add_equivalence(
+        U2Gate(phi, lam), def_sy_u2)
+
+
+You will want this to be run on import so that as soon as the provider's
+package is imported it will be run. This will ensure that any time the
+:class:`~qiskit.transpiler.passes.BasisTranslator` pass is run with the
+custom gates the equivalence rules are defined.
+
+It's also worth noting that depending on the basis your using some optimization
+passes in the transpiler, such as
+:class:`~qiskit.transpiler.passes.Optimize1qGatesDecomposition`, may not
+be able to operate with your custom basis. For example continuing with the
+``SYGate`` example the
+:class:`~qiskit.transpiler.passes.Optimize1qGatesDecomposition` will not be
+able to simplify runs of single qubit gates into the SY basis. This is because
+the :class:`~qiskit.quantum_info.OneQubitEulerDecomposer` class does not
+know how to work in the SY basis. To solve this the ``SYGate`` class would
+need to be added to Qiskit and
+:class:`~qiskit.quantum_info.OneQubitEulerDecomposer` updated to support
+decomposing to the ``SYGate``. Longer term that is likely a better direction
```

Comment:
Is this info documented in the transpiler? It would be nicer here to be able to say "Some things won't work for any basis, if you run into this problem, look at the transpiler module docs"

Reply:
It doesn't look like it, looking at the docs for the pass itself: https://github.com/Qiskit/qiskit-terra/blob/master/qiskit/transpiler/passes/optimization/optimize_1q_decomposition.py there is not really any detail in there (which is my fault). 

The transpile documentation doesn't really go into this level of detail either:

- https://github.com/Qiskit/qiskit-terra/blob/master/qiskit/transpiler/__init__.py is pretty high level about the theory of operation
-  https://github.com/Qiskit/qiskit-terra/blob/master/qiskit/compiler/transpiler.py is about the mechanics of the api for calling `transpile`

The one place that has this documented is: https://github.com/Qiskit/qiskit-terra/blob/master/qiskit/quantum_info/synthesis/one_qubit_decompose.py but you'd need to know that the `Optimize1qGatesDecomposition`  pass is using that internally.